trap-handlers.adoc

```diff
@@ -0,0 +1,127 @@
+[[trap-handlers]]
+== Machine-Level Interrupts and Exceptions
+
+=== Delegation
+
+A RISC-V Machine mode execution environment may delegate handling interrupts and exceptions to the Supervisor or Hypervisor mode execution environment.
+
+For reference, see https://github.com/riscv/riscv-isa-manual/blob/main/src/machine.adoc#machine-cause-register-mcause[the definitions in the RISC-V ISA spec].
+
+==== Interrupts
+[%autowidth,float="center",align="center",cols=">,>,<",options="header",]
+|===
+| Code / Description |OpenSBI |oreboot SBI
```

Comment:
Why reference m-mode implementations? Why not call out what the expected behavior should be? And we should probably include the reasoning.

Reply:
this is for reference now, based on this suggestion https://github.com/riscv-non-isa/riscv-brs/issues/83#issuecomment-1764583868

my pledge is to always delegate everything that S-mode could handle, which only excludes operations involving CSRs that are M-mode only, AIUI

Reply:
If I had implied we should go into describing M mode implementations, that was not the intention. The BRS spec describes S mode behavior as it relates to booting at OS. Some implementations may not even have an M mode (think VMs).

We should come up with expected behavior (or a mechanism of conveying the necessary info). But I kinda wanna second Ved's feedback or why we ought to care - I think the assumption is that the OS will take the exception eventually for reasons that it would care about (even if the original exception went to M, as it very well might, for example due to firwmare-first error handling)

Reply:
"my pledge is to always delegate everything that S-mode could handle" we're definitely into impl-specific choices, and I don't think the BRS spec is the place to push for those choices. Like with other architectures, work-arounds and additional complexity may end up getting put into M, but in a manner generally transparent to S mode software. What matters is not what's done, but what's the measurable impact.

Reply:
Gotcha. I had understood from this comment that BRS would supersede the platform spec:
https://github.com/riscv/riscv-platform-specs/pull/91#issuecomment-1707916908

So my intention was to get to a definition of behavior an S-mode OS would expect to avoid confusion where exceptions and/or interrupts are or are not arriving. That is what surprised me when Linux was not handling unaligned access traps, which is now being implemented in a pending patch series. Other OSes that are not part of that development history would benefit. Should this rather go into PRS, then? I would be happy to see it well-defined _somewhere_, not being picky where. BRS appeared to be closest to that from my current understanding. Where do you rather see it? It could also live in the wiki or in a blog post, I think.

For clarification: In which sense do you mean "transparent"?
There are two interpretations around:
1. opaque, the other component (S-mode here) is not aware of something
2. visible, the other component can notice and acknowledge

Reply:
Got it, so basically you want a base set of exceptions that an S-Mode OS can expect to be delivered? (like unaligned access).

Will bring this up in the BRS call today and come back with some guidance.